File: examples/typescript/codec-vectors.test.ts
import { describe, it, expect } from 'vitest';
import vectors from '@moqtap/test-vectors/transport/draft14/codec/messages/subscribe.json';
import { decodeMessage, encodeMessage } from '@moqtap/codec';

function hexToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substring(i, i + 2), 16);
  }
  return bytes;
}

function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
}

describe(`codec: ${vectors.message_type}`, () => {
  for (const v of vectors.vectors) {
    it(v.description, () => {
      const bytes = hexToBytes(v.hex);
      const isCanonical = v.canonical ?? true;

      if (v.decoded) {
        // Decode direction: always test
        const result = decodeMessage(bytes);
        expect(result).toEqual(v.decoded);

        // Encode direction: only for canonical vectors
        if (isCanonical) {
          const reEncoded = encodeMessage(result);
          expect(bytesToHex(reEncoded)).toBe(v.hex);
        }
      } else if (v.error) {
        expect(() => decodeMessage(bytes)).toThrow();
      }
    });
  }
});

File: examples/go/codec_vectors_test.go
package vectors_test

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"testing"
)

type VectorFile struct {
	MessageType string   `json:"message_type"`
	Vectors     []Vector `json:"vectors"`
}

type Vector struct {
	Description string           `json:"description"`
	Hex         string           `json:"hex"`
	Decoded     *json.RawMessage `json:"decoded,omitempty"`
	Error       *string          `json:"error,omitempty"`
	Canonical   *bool            `json:"canonical,omitempty"`
}

func TestSubscribeVectors(t *testing.T) {
	data, err := os.ReadFile(
		"test-vectors/transport/draft14/codec/messages/subscribe.json",
	)
	if err != nil {
		t.Fatal("vector file not found — did you init the submodule?")
	}
	var file VectorFile
	json.Unmarshal(data, &file)

	for _, v := range file.Vectors {
		t.Run(v.Description, func(t *testing.T) {
			bytes, _ := hex.DecodeString(v.Hex)
			isCanonical := v.Canonical == nil || *v.Canonical

			if v.Decoded != nil {
				// Decode direction: always test
				_ = bytes // result := yourlib.DecodeMessage(bytes)
				if isCanonical {
					// Encode direction: only for canonical vectors
					// reEncoded := yourlib.EncodeMessage(result)
				}
			} else if v.Error != nil {
				// Invalid case: expect error
				_ = bytes // _, err := yourlib.DecodeMessage(bytes)
			}
		})
	}
}
